orttraining/orttraining/lazy_tensor/accelerator.cpp

```diff
@@ -0,0 +1,488 @@
+
+  // Run the compiled function!
+  auto outputs = cache_[spec].code(inputs);
+
+  // Discard used inputs.
+  torch::jit::drop(stack, inputs.size());
+
+  // Return results to caller.
+  for (auto& output : outputs) {
+    stack.push_back(output);
+  }
+
+  if (DumpInputsOutputs()) {
+    at::ArrayRef<c10::IValue> outputs = torch::jit::last(stack, subgraph_->outputs().size());
+    std::cout << "[ORT,Output] " << ToString(outputs) << std::endl;
+  }
+}
+
+void Accelerator::PytorchRun(torch::jit::Stack& stack) {
+  DynamicSettings::GetInstance().SetOnnxFusionFlag(false);
+#ifdef USE_CUDA
+  NvtxRange range(__func__);
+#endif
+  if (DumpGraph()) {
+    std::cout << "[Pytorch,Graph]\n"
+              << subgraph_->toString(true);
+  }
+  if (DumpInputsOutputs()) {
+    at::ArrayRef<c10::IValue> inputs = torch::jit::last(
+        stack, subgraph_->inputs().size());
+    std::cout << "[PyTorch,Input] " << ToString(inputs) << std::endl;
+  }
+
+  torch::jit::GraphExecutor executor(subgraph_, "");
+  executor.run(stack);
+
+  if (DumpInputsOutputs()) {
+    at::ArrayRef<c10::IValue> outputs = torch::jit::last(
+        stack, subgraph_->outputs().size());
+    std::cout << "[PyTorch,Output] " << ToString(outputs) << std::endl;
+  }
+  DynamicSettings::GetInstance().SetOnnxFusionFlag(true);
+}
+
+void Accelerator::DebugRun(torch::jit::Stack& stack) {
+#ifdef USE_CUDA
+  NvtxRange range(__func__);
+#endif
+  torch::jit::Stack copy;
+  copy = stack;
+  OrtRun(stack);
+  PytorchRun(copy);
+  ORT_ENFORCE(CompareStack(stack, copy),
+              "ORT and Pytorch must generate the same results "
+              "but tensor types, shapes or content are different. "
+              "Use, e.g., ORT_LT_RELATIVE_TOLERANCE=1e-3 and "
+              "ORT_LT_ABSOLUTE_TOLERANCE=1e-4 "
+              "to increase the content tolerance, if "
+              "the difference is due to numerical errors.");
+}
+
+void Accelerator::Run(torch::jit::Stack& stack) {
+  const auto run_type = RunType();
+  if (run_type == "debug") {
+    // Run both ORT and Pytorch to execute the subgraph
+    // and compare their output types and shapes.
+    DebugRun(stack);
+  } else if (run_type == "ort") {
+    OrtRun(stack);
+  } else if (run_type == "pytorch") {
+    PytorchRun(stack);
+  } else {
+    ORT_THROW("Unknown run type: ", run_type);
+  }
+}
+
+static void CheckArgs(
+    const at::ArrayRef<c10::IValue>& inputs) {
+  // TODO: remove this check.
+#ifdef USE_CUDA
+  NvtxRange range(__func__);
+#endif
+  TORCH_CHECK(inputs.size(), "Need at least one input.");
+  for (const auto& input : inputs) {
+    TORCH_CHECK(input.isTensor() || input.isScalar(), "Compiler can only handle Tensor or Scalar inputs.");
+  }
+}
+
+// Store input types in sub-graph so that
+// ONNX exporter can use them. Input types
+// are required when executing ONNX model
+// in ORT.
+// TODO: Allow ORT to accept models without
+// input types. Then, we can remove this function.
+static void PropagateArgTypes(
```

Comment:
is this method just set the types for graph inputs? how is it related with "Propagate"?

Reply:
I will rename it to `SetArgTypes`. Thanks.